src/launcherext.ts

```diff
@@ -0,0 +1,169 @@
+     * @returns An array of tuples containing string to display
+     */
+    search_results?: (text: string) => Array<SearchResult> | null;
+}
+
+export class CalcLauncher implements LauncherExtension {
+    prefix = '=';
+    name = 'calc';
+    ext?: Ext;
+    search?: Search;
+
+    init(ext: Ext, search: Search): this {
+        this.ext = ext;
+        this.search = search;
+
+        return this;
+    }
+
+    apply = (expr: string): boolean => {
+        const value: string = evaluate(expr).toString();
+        log.info(`${expr} = ${value}`);
+        if (!this.search) {
+            log.error("init was never called");
+        }
+
+        this.search?.set_text(`=${value}`);
+
+        return true;
+    }
+
+    search_results = (expr: string): Array<SearchResult> | null => {
+        const item: SearchResult =
+        {
+            display: `=${evaluate(expr).toString()}`,
+            category_icon: new St.Icon({
+                icon_name: 'x-office-spreadsheet', // looks like calculations?
+                icon_size: LAUNCHER_ICON_SIZE / 2,
+                style_class: "pop-shell-search-cat"
+            }),
+            app_icon: new St.Icon({
+                icon_name: 'accessories-calculator',
+                icon_size: LAUNCHER_ICON_SIZE
+            })
+        };
+
+        return [item];
+    }
+}
+
+export class CommandLauncher implements LauncherExtension {
+    prefix = ':';
+    name = 'command';
+
+    init(): this {
+        return this;
+    }
+
+    apply(cmd: string): boolean {
+        spawnCommandLine(cmd);
+        return false;
+    }
+}
+
+export class TerminalLauncher implements LauncherExtension {
+    prefix = 't:';
+    name = 'terminal';
+
+    init(): this {
+        return this;
+    }
+
+    apply(cmd: string): boolean {
+        let terminal = TERMINAL.get_or_init(() => {
+            let path: string | null = GLib.find_program_in_path('x-terminal-emulator');
+            return path ?? 'gnome-terminal';
+        });
+
+        spawnCommandLine(`${terminal} -e sh -c '${cmd}; echo "Press to exit"; read t'`);
+        return false;
+    }
+}
+
+export class WebSearchLauncher implements LauncherExtension {
+    prefix = '?';
+    name = 'web-search';
+    ext?: Ext;
+    search?: Search;
+
+    init(ext: Ext, search: Search): this {
+        this.ext = ext;
+        this.search = search;
+
+        return this;
+    }
+
+    private get_query(webSearch: string): string {
+        const searchBase = this.ext?.settings.search_engine();
+        log.info(searchBase ?? 'search engine undefined');
+        return searchBase + encodeURIComponent(webSearch);
+    }
+
+    apply(webSearch: string): boolean {
+        // xdg-open should use user's default browser
+        const cmd = `xdg-open "${this.get_query(webSearch)}"`;
+        spawnCommandLine(cmd);
+
+        return false;
+    }
+
+    search_results(webSearch: string): Array<SearchResult> | null {
```

Comment:
Not sure what the icons should be here.
![image](https://user-images.githubusercontent.com/661373/81134983-7f909d80-8f24-11ea-8743-cc08404265f7.png)

Reply:
I wonder if there would be value in display the search results of a query here, instead of the URL of the query.

Reply:
I think results would be an awesome enhancement. Being able to select the item and launch the browser to the page instead of just launching the query. I think it's a bit more complicated to figure out though since it starts adding web calls from launcher, and having a different design for web results.

Reply:
I was looking into having a web view be displayed, but it looks like it's not supported by Clutter. DuckDuckGo has an [instant answers api](https://duckduckgo.com/api), but it's not the same as full search results.

Reply:
It's not, but you could scrape the HTML output for titles & links and list search results that way

Reply:
I started looking into this, I can get the HTML using Soup, but there doesn't seem to be a decent XML Parser in GJS? It looks like there's [Rest.XmlParser](https://gjs-docs.gnome.org/rest07~0.7_api/rest.xmlparser#method-parse_from_data), but playing around in GJS, I can only see the root HTML node lol. I think we'd also have to add support for async search results in search since doing synchronous sessions would pause user typing I think.

Reply:
If it's a lot of work, it'd be best to wait for GJS to have support. I'm impatiently waiting for the day we can use Rust WebASM modules in GJS.